### PR TITLE
Cleanup: scheduler tests independent of job controller

### LINKED ISSUE #155: Scheduler integration tests should be independent of the job controller
**What would you like to be cleaned**:

The scheduler integration should create QueuedWorkloads instead of Jobs.

**Why is this needed**:

Separation of concerns.

### PR DESCRIPTION
**Pull request by thisisprasad**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it:**
Made scheduler integration tests independent of job controller

**Which issue(s) this PR fixes:**
Fixes #155 

**Special notes for your reviewer:**

### CODE REVIEW

**Comment by kerthcet** [comment]
I think you want to run the tests.
/ok-to-test

**Comment by alculquicondor** [comment]
is this still work in progress or ready to review?

**Comment by thisisprasad** [comment]
> is this still work in progress or ready to review?

Yes, work is still in-progress, the job-controller tests are failing on my local minikube cluster.
For scheduler integration tests, the webhook check is failing. The cluster requires `v1beta1` as the `admissionreviewversion` as well.

I have updated the code now. Will monitor the tests and post an update.

**Comment by thisisprasad** [comment]
@alculquicondor @kerthcet the PR is ready for review.

File: test/integration/controller/job/job_controller_test.go [hunk 1]
```diff
@@ -275,3 +276,209 @@ var _ = ginkgo.Describe("Job controller for workloads with no queue set", func()
 		}, framework.Timeout, framework.Interval).Should(gomega.Succeed())
 	})
 })
+
+var _ = ginkgo.Describe("Job controller for scheduling jobs on clusterQueue", ginkgo.Ordered, func() {
```
Comment (alculquicondor) [inline test/integration/controller/job/job_controller_test.go:280]:
```suggestion
var _ = ginkgo.Describe("Job controller interacting with scheduler", ginkgo.Ordered, func() {
```

File: test/integration/framework/framework.go [hunk 2]
```diff
@@ -256,6 +257,15 @@ func ExpectWorkloadsToBeFrozen(ctx context.Context, k8sClient client.Client, cq
 	}, Timeout, Interval).Should(gomega.Equal(len(wls)), "Not enough workloads are frozen")
 }
 
+func ExpectWorkloadToBeAdmittedAs(ctx context.Context, k8sClient client.Client, wlNsName types.NamespacedName,
+	admission *kueue.Admission) {
```
Comment (alculquicondor) [inline test/integration/framework/framework.go:261]:
```suggestion
func ExpectWorkloadToBeAdmittedAs(ctx context.Context, k8sClient client.Client, wlNsName types.NamespacedName, admission *kueue.Admission) {
```

File: test/integration/scheduler/scheduler_test.go [hunk 3]
```diff
@@ -74,166 +73,6 @@ var _ = ginkgo.Describe("Scheduler", func() {
...
-		ginkgo.AfterEach(func() {
-			gomega.Expect(framework.DeleteNamespace(ctx, k8sClient, ns)).To(gomega.Succeed())
-			framework.ExpectClusterQueueToBeDeleted(ctx, k8sClient, prodClusterQ, true)
-			framework.ExpectClusterQueueToBeDeleted(ctx, k8sClient, devClusterQ, true)
-			framework.ExpectResourceFlavorToBeDeleted(ctx, k8sClient, onDemandFlavor, true)
-			gomega.Expect(framework.DeleteResourceFlavor(ctx, k8sClient, spotTaintedFlavor)).To(gomega.Succeed())
-			gomega.Expect(framework.DeleteResourceFlavor(ctx, k8sClient, spotUntaintedFlavor)).To(gomega.Succeed())
-		})
-
-		ginkgo.It("Should schedule jobs as they fit in their ClusterQueue", func() {
```
Comment (alculquicondor) [inline test/integration/scheduler/scheduler_test.go:123]: We should have a version of this test with just workloads.

File: test/integration/controller/job/job_controller_test.go [hunk 4]
```diff
@@ -275,3 +276,209 @@ var _ = ginkgo.Describe("Job controller for workloads with no queue set", func()
...
+			Taint(corev1.Taint{
+				Key:    instanceKey,
+				Value:  "spot-tainted",
+				Effect: corev1.TaintEffectNoSchedule,
+			}).Obj()
+
+		spotUntaintedFlavor = testing.MakeResourceFlavor("spot-untainted").Label(instanceKey, "spot-untainted").Obj()
+	})
+
+	ginkgo.When("Scheduling v1.Job", func() {
```
Comment (alculquicondor) [inline test/integration/controller/job/job_controller_test.go:325]: We don't need this extra `When`

File: test/integration/controller/job/job_controller_test.go [hunk 5]
```diff
@@ -275,3 +276,209 @@ var _ = ginkgo.Describe("Job controller for workloads with no queue set", func()
...
+		ginkgo.AfterEach(func() {
+			gomega.Expect(framework.DeleteNamespace(ctx, k8sClient, ns)).To(gomega.Succeed())
+			framework.ExpectClusterQueueToBeDeleted(ctx, k8sClient, prodClusterQ, true)
+			framework.ExpectClusterQueueToBeDeleted(ctx, k8sClient, devClusterQ, true)
+			framework.ExpectResourceFlavorToBeDeleted(ctx, k8sClient, onDemandFlavor, true)
+			gomega.Expect(framework.DeleteResourceFlavor(ctx, k8sClient, spotTaintedFlavor)).To(gomega.Succeed())
+			gomega.Expect(framework.DeleteResourceFlavor(ctx, k8sClient, spotUntaintedFlavor)).To(gomega.Succeed())
+		})
+
+		ginkgo.It("Should schedule jobs as they fit in their ClusterQueue", func() {
```
Comment (alculquicondor) [inline test/integration/controller/job/job_controller_test.go:371]: We probably need just this test.

File: test/integration/scheduler/scheduler_test.go [hunk 6]
```diff
@@ -74,166 +73,6 @@ var _ = ginkgo.Describe("Scheduler", func() {
...
-			gomega.Eventually(func() *bool {
-				gomega.Expect(k8sClient.Get(ctx, lookupKey2, createdProdJob2)).Should(gomega.Succeed())
-				return createdProdJob2.Spec.Suspend
-			}, framework.Timeout, framework.Interval).Should(gomega.Equal(pointer.Bool(false)))
-			gomega.Expect(createdProdJob2.Spec.Template.Spec.NodeSelector[instanceKey]).Should(gomega.Equal(onDemandFlavor.Name))
-			framework.ExpectPendingWorkloadsMetric(prodClusterQ, 0)
-			framework.ExpectAdmittedActiveWorkloadsMetric(prodClusterQ, 1)
-		})
-
-		ginkgo.It("Should schedule jobs according to their priorities", func() {
```
Comment (alculquicondor) [inline test/integration/scheduler/scheduler_test.go:180]: This one too

File: pkg/util/testing/wrappers.go [hunk 7]
```diff
@@ -214,6 +215,11 @@ func (w *WorkloadWrapper) Toleration(t corev1.Toleration) *WorkloadWrapper {
 	return w
 }
 
+func (w *WorkloadWrapper) NodeSelector(k, v string) *WorkloadWrapper {
```
Comment (alculquicondor) [inline pkg/util/testing/wrappers.go:218]: nit: pass the whole map.

File: test/integration/framework/framework.go (hunk 2 above)
Comment (alculquicondor) [reply test/integration/framework/framework.go:261]: Also, any reason not to accept a `kueue.Workload` like `ExpectWorkloadToBeAdmitted`?

File: test/integration/scheduler/scheduler_test.go [hunk 8]
```diff
@@ -337,60 +147,38 @@ var _ = ginkgo.Describe("Scheduler", func() {
...
-			createdJob1 := &batchv1.Job{}
-			gomega.Eventually(func() *bool {
-				lookupKey := types.NamespacedName{Name: job1.Name, Namespace: job1.Namespace}
-				gomega.Expect(k8sClient.Get(ctx, lookupKey, createdJob1)).Should(gomega.Succeed())
-				return createdJob1.Spec.Suspend
-			}, framework.Timeout, framework.Interval).Should(gomega.Equal(pointer.Bool(false)))
-			gomega.Expect(createdJob1.Spec.Template.Spec.NodeSelector[instanceKey]).Should(gomega.Equal(onDemandFlavor.Name))
+			wl1 := testing.MakeWorkload("on-demand-wl1", ns.Name).Queue(fooQ.Name).Request(corev1.ResourceCPU, "8").Obj()
+			gomega.Expect(k8sClient.Create(ctx, wl1)).Should(gomega.Succeed())
+			framework.ExpectWorkloadsToBeAdmitted(ctx, k8sClient, fooCQ.Name, wl1)
```
Comment (alculquicondor) [inline test/integration/scheduler/scheduler_test.go:152]: this line seems redundant.

File: test/integration/scheduler/scheduler_test.go [hunk 9]
```diff
@@ -337,60 +147,38 @@ var _ = ginkgo.Describe("Scheduler", func() {
...
-			createdJob3 := &batchv1.Job{}
-			gomega.Eventually(func() *bool {
-				lookupKey := types.NamespacedName{Name: job3.Name, Namespace: job3.Namespace}
-				gomega.Expect(k8sClient.Get(ctx, lookupKey, createdJob3)).Should(gomega.Succeed())
-				return createdJob3.Spec.Suspend
-			}, framework.Timeout, framework.Interval).Should(gomega.Equal(pointer.Bool(false)))
-			gomega.Expect(createdJob3.Spec.Template.Spec.NodeSelector[instanceKey]).Should(gomega.Equal(onDemandFlavor.Name))
+			wl3 := testing.MakeWorkload("on-demand-wl3", ns.Name).Queue(fooQ.Name).Request(corev1.ResourceCPU, "2").Obj()
+			gomega.Expect(k8sClient.Create(ctx, wl3)).Should(gomega.Succeed())
+			framework.ExpectWorkloadsToBeAdmitted(ctx, k8sClient, fooCQ.Name, wl3)
```
Comment (alculquicondor) [inline test/integration/scheduler/scheduler_test.go:169]: also redundant

File: test/integration/scheduler/scheduler_test.go [hunk 10]
```diff
@@ -337,60 +147,38 @@ var _ = ginkgo.Describe("Scheduler", func() {
...
-				}})).Should(gomega.Succeed())
-
-			gomega.Eventually(func() *bool {
-				lookupKey := types.NamespacedName{Name: job2.Name, Namespace: job2.Namespace}
-				gomega.Expect(k8sClient.Get(ctx, lookupKey, createdJob2)).Should(gomega.Succeed())
-				return createdJob2.Spec.Suspend
-			}, framework.Timeout, framework.Interval).Should(gomega.Equal(pointer.Bool(false)))
-			gomega.Expect(createdJob2.Spec.Template.Spec.NodeSelector[instanceKey]).Should(gomega.Equal(onDemandFlavor.Name))
+			gomega.Expect(k8sClient.Delete(ctx, wl1, client.PropagationPolicy(metav1.DeletePropagationBackground))).Should(gomega.Succeed())
+			framework.ExpectWorkloadsToBeAdmitted(ctx, k8sClient, cq.Name, wl2)
```
Comment (alculquicondor) [inline test/integration/scheduler/scheduler_test.go:178]: redundant

File: test/integration/scheduler/scheduler_test.go [hunk 11]
```diff
@@ -440,12 +223,10 @@ var _ = ginkgo.Describe("Scheduler", func() {
 			updatedCq.Spec.Resources = []kueue.Resource{*updatedResource}
 			gomega.Expect(k8sClient.Update(ctx, updatedCq)).Should(gomega.Succeed())
 
-			gomega.Eventually(func() *bool {
-				lookupKey := types.NamespacedName{Name: job.Name, Namespace: job.Namespace}
-				gomega.Expect(k8sClient.Get(ctx, lookupKey, createdJob2)).Should(gomega.Succeed())
-				return createdJob2.Spec.Suspend
-			}, framework.Timeout, framework.Interval).Should(gomega.Equal(pointer.Bool(false)))
-			gomega.Expect(createdJob2.Spec.Template.Spec.NodeSelector[instanceKey]).Should(gomega.Equal(onDemandFlavor.Name))
+			framework.ExpectWorkloadsToBeAdmitted(ctx, k8sClient, cq.Name, wl)
```
Comment (alculquicondor) [inline test/integration/scheduler/scheduler_test.go:226]: redundant

File: test/integration/scheduler/scheduler_test.go [hunk 12]
```diff
@@ -495,35 +276,22 @@ var _ = ginkgo.Describe("Scheduler", func() {
...
-			gomega.Expect(k8sClient.Create(ctx, job2)).Should(gomega.Succeed())
-			createdJob2 := &batchv1.Job{}
-			gomega.Eventually(func() *bool {
-				gomega.Expect(k8sClient.Get(ctx, types.NamespacedName{Name: job2.Name, Namespace: job2.Namespace}, createdJob2)).Should(gomega.Succeed())
-				return createdJob2.Spec.Suspend
-			}, framework.Timeout, framework.Interval).Should(gomega.Equal(pointer.Bool(true)), "Job should be suspended")
-
-			framework.ExpectPendingWorkloadsMetric(cq, 2)
+		ginkgo.It("Should admit workloads from the selected namespaces", func() {
+			ginkgo.By("checking a workload does not get admitted at first")
```
Comment (alculquicondor) [inline test/integration/scheduler/scheduler_test.go:280]: The test had two workloads. Can you do the same?

File: test/integration/scheduler/scheduler_test.go [hunk 13]
```diff
@@ -495,35 +276,22 @@ var _ = ginkgo.Describe("Scheduler", func() {
...
 			ns.Labels = map[string]string{"dep": "eng"}
 			gomega.Expect(k8sClient.Update(ctx, ns)).Should(gomega.Succeed())
-			gomega.Eventually(func() *bool {
-				gomega.Expect(k8sClient.Get(ctx, types.NamespacedName{Name: job1.Name, Namespace: job1.Namespace}, createdJob1)).Should(gomega.Succeed())
-				return createdJob1.Spec.Suspend
-			}, framework.Timeout, framework.Interval).Should(gomega.Equal(pointer.Bool(false)), "Job should be unsuspended")
-			framework.ExpectPendingWorkloadsMetric(cq, 1)
+			framework.ExpectWorkloadsToBeAdmitted(ctx, k8sClient, cq.Name, wl)
+			lookupWlKey := types.NamespacedName{Name: wl.Name, Namespace: wl.Namespace}
+			expectAdmission := testing.MakeAdmission(cq.Name).Flavor(corev1.ResourceCPU, onDemandFlavor.Name).Obj()
```
Comment (alculquicondor) [inline test/integration/scheduler/scheduler_test.go:292]: I don't think we care about the flavors in this test. I would remove this just to keep the test simple.

File: test/integration/scheduler/scheduler_test.go [hunk 14]
```diff
@@ -666,32 +434,30 @@ var _ = ginkgo.Describe("Scheduler", func() {
...
-				return createdJob1.Spec.Suspend
-			}, framework.Timeout, framework.Interval).Should(gomega.Equal(pointer.Bool(false)))
-			gomega.Expect(createdJob1.Spec.Template.Spec.NodeSelector[instanceKey]).Should(gomega.Equal(spotUntaintedFlavor.Name))
-			framework.ExpectPendingWorkloadsMetric(cq, 0)
+			ginkgo.By("checking a workload without affinity gets admitted on the first flavor")
+			wl1 := testing.MakeWorkload("no-affinity-workload", ns.Name).Queue(queue.Name).Request(corev1.ResourceCPU, "1").Obj()
+			gomega.Expect(k8sClient.Create(ctx, wl1)).Should(gomega.Succeed())
+			expectAdmission := testing.MakeAdmission(cq.Name).Flavor(corev1.ResourceCPU, spotUntaintedFlavor.Name).Obj()
+			wlCopy := &kueue.Workload{}
+			gomega.Eventually(func() *kueue.Admission {
```
Comment (alculquicondor) [inline test/integration/scheduler/scheduler_test.go:442]: Use `ExpectWorkloadToBeAdmittedAs`?

File: test/integration/scheduler/scheduler_test.go [hunk 15]
```diff
@@ -666,32 +434,30 @@ var _ = ginkgo.Describe("Scheduler", func() {
...
-			createdJob2 := &batchv1.Job{}
-			gomega.Eventually(func() bool {
-				lookupKey := types.NamespacedName{Name: job2.Name, Namespace: job2.Namespace}
-				return k8sClient.Get(ctx, lookupKey, createdJob2) == nil && !*createdJob2.Spec.Suspend
-			}, framework.Timeout, framework.Interval).Should(gomega.BeTrue())
-			gomega.Expect(len(createdJob2.Spec.Template.Spec.NodeSelector)).Should(gomega.Equal(2))
-			gomega.Expect(createdJob2.Spec.Template.Spec.NodeSelector[instanceKey]).Should(gomega.Equal(onDemandFlavor.Name))
+			gomega.Expect(k8sClient.Create(ctx, wl2)).Should(gomega.Succeed())
+			expectAdmission = testing.MakeAdmission(cq.Name).Flavor(corev1.ResourceCPU, onDemandFlavor.Name).Obj()
+			gomega.Eventually(func() *kueue.Admission {
```
Comment (alculquicondor) [inline test/integration/scheduler/scheduler_test.go:456]: `ExpectWorkloadToBeAdmittedAs`

File: test/integration/scheduler/scheduler_test.go [hunk 16]
```diff
@@ -755,11 +516,7 @@ var _ = ginkgo.Describe("Scheduler", func() {
 				gomega.Expect(framework.DeleteClusterQueue(ctx, k8sClient, fallbackClusterQueue)).ToNot(gomega.HaveOccurred())
 			}()
 
-			gomega.Eventually(func() *bool {
-				gomega.Expect(k8sClient.Get(ctx, lookupKey, createdJob)).Should(gomega.Succeed())
-				return createdJob.Spec.Suspend
-			}, framework.Timeout, framework.Interval).Should(gomega.Equal(pointer.Bool(false)))
-			gomega.Expect(createdJob.Spec.Template.Spec.NodeSelector[instanceKey]).Should(gomega.Equal(onDemandFlavor.Name))
+			framework.ExpectWorkloadsToBeAdmitted(ctx, k8sClient, prodBEClusterQ.Name, wl)
```
Comment (alculquicondor) [inline test/integration/scheduler/scheduler_test.go:519]: In this case the flavor was important. Use `ExpectWorkloadToBeAdmittedAs`

File: test/integration/scheduler/scheduler_test.go [hunk 17]
```diff
@@ -868,8 +625,10 @@ var _ = ginkgo.Describe("Scheduler", func() {
 			ginkgo.By("Marking the big workload as finished")
 			framework.UpdateWorkloadStatus(ctx, k8sClient, bigWl, func(wl *kueue.Workload) {
 				wl.Status.Conditions = append(wl.Status.Conditions, kueue.WorkloadCondition{
-					Type:   kueue.WorkloadFinished,
-					Status: corev1.ConditionTrue,
+					Type:               kueue.WorkloadFinished,
+					Status:             corev1.ConditionTrue,
+					LastProbeTime:      metav1.Now(),
+					LastTransitionTime: metav1.Now(),
```
Comment (alculquicondor) [inline test/integration/scheduler/scheduler_test.go:631]: Is this necessary?

**Comment by ahg-g** [comment]
@thisisprasad  friendly ping, do you mind addressing the comments and move forward with this open PR?

**Comment by thisisprasad** [comment]
2 scheduler integration tests are failing. Debugging the issue. Marking the PR as WIP.

**Comment by thisisprasad** [comment]
The It `Should admit workloads according to their priorities` is flaky test and is failing intermittently. The test is expecting a low priority workload be in `pending` state, i.e workload should be reconciled once and has pending condition in it's status. But that's not happening everytime.

I have tried increasing the field check eventual timeout to approx. 60 sec and interval to 1~2 secs. But no luck :(

Can anyone please suggest an alternative approach to write the test? And what will be the scope of the changes required.

File: test/integration/framework/framework.go (hunk 2 above)
Comment (thisisprasad) [reply test/integration/framework/framework.go]: Added *kueue.Workload as function argument. Currently the func derives key from object only.
Earlier we were passing NamespacedName key.

File: test/integration/scheduler/scheduler_test.go (hunk 3 above)
Comment (thisisprasad) [reply test/integration/scheduler/scheduler_test.go:123]: Added

File: test/integration/controller/job/job_controller_test.go (hunk 4 above)
Comment (thisisprasad) [reply test/integration/controller/job/job_controller_test.go]: Removed

File: test/integration/controller/job/job_controller_test.go (hunk 5 above)
Comment (thisisprasad) [reply test/integration/controller/job/job_controller_test.go]: Added

File: test/integration/scheduler/scheduler_test.go (hunk 6 above)
Comment (thisisprasad) [reply test/integration/scheduler/scheduler_test.go:180]: Added

File: pkg/util/testing/wrappers.go (hunk 7 above)
Comment (thisisprasad) [reply pkg/util/testing/wrappers.go]: Whole map is being passed now

File: test/integration/scheduler/scheduler_test.go (hunk 12 above)
Comment (thisisprasad) [reply test/integration/scheduler/scheduler_test.go]: Added. Refactored the tests with two workloads.

File: test/integration/scheduler/scheduler_test.go (hunk 17 above)
Comment (thisisprasad) [reply test/integration/scheduler/scheduler_test.go]: Removed

**Comment by thisisprasad** [comment]
> The It `Should admit workloads according to their priorities` is flaky test and is failing intermittently. The test is expecting a low priority workload be in `pending` state, i.e workload should be reconciled once and has pending condition in it's status. But that's not happening everytime.
> 
> I have tried increasing the field check eventual timeout to approx. 60 sec and interval to 1~2 secs. But no luck :(
> 
> Can anyone please suggest an alternative approach to write the test? And what will be the scope of the changes required.

@alculquicondor @ahg-g any suggestions on the above issue?

**Comment by alculquicondor** [comment]
Is it flaky in the `main` branch?

**Comment by thisisprasad** [comment]
Not it is not flaky in `main` branch. But, the `main` branch has the job version of this test.

**Comment by alculquicondor** [comment]
@kerthcet have you observed anything like this?

**Comment by kerthcet** [comment]
This is because we used to generate workloads via job controller, then the priority of the workload will be updated according to the priorityClassName, but now, we create the workloads via `MakeWorkload()` and we don't specify the priority values, If so,  the priority value will always be 0, so the earlier created one will be located at the top of the heap and pop out first.
```
func Priority(w *kueue.Workload) int32 {
	if w.Spec.Priority != nil {
		return *w.Spec.Priority
	}
	// When priority of a running workload is nil, it means it was created at a time
	// that there was no global default priority class and the priority class
	// name of the pod was empty. So, we resolve to the static default priority.
	return constants.DefaultPriority
}
```

What we want to do is just change like below, if so, `PriorityClass` creation seems longer needed.
```
wlLowPriority := testing.MakeWorkload().PriorityClass(lowPriorityClass.Name).Obj()
```
To:
```
wlLowPriority := testing.MakeWorkload().Priority(priority).Obj()
```

**Comment by kerthcet** [comment]
I think we should add some validation to the workload.
When creating/updating workloads: 
- [ ] If priorityClassName != "", the priority shouldn't be nil.

File: test/integration/controller/job/job_controller_test.go [hunk 18]
```diff
@@ -275,3 +276,141 @@ var _ = ginkgo.Describe("Job controller for workloads with no queue set", func()
...
+		spotTaintedFlavor   *kueue.ResourceFlavor
+		spotUntaintedFlavor *kueue.ResourceFlavor
+		prodClusterQ        *kueue.ClusterQueue
+		devClusterQ         *kueue.ClusterQueue
+		prodQueue           *kueue.Queue
+		devQueue            *kueue.Queue
+	)
+
+	ginkgo.BeforeEach(func() {
+		fwk = &framework.Framework{
```
Comment (kerthcet) [inline test/integration/controller/job/job_controller_test.go:297]: We can place this in the `suite_test.go`, refer to `test/integration/controller/core/suite_test.go`

File: test/integration/scheduler/scheduler_test.go [hunk 19]
```diff
@@ -265,58 +217,29 @@ var _ = ginkgo.Describe("Scheduler", func() {
...
-					Namespace: job1.Namespace,
-				}})).Should(gomega.Succeed())
-			gomega.Eventually(func() *bool {
-				lookupKey := types.NamespacedName{Name: job2.Name, Namespace: job2.Namespace}
-				gomega.Expect(k8sClient.Get(ctx, lookupKey, createdJob2)).Should(gomega.Succeed())
-				return createdJob2.Spec.Suspend
-			}, framework.Timeout, framework.Interval).Should(gomega.Equal(pointer.Bool(false)))
-			gomega.Expect(createdJob2.Spec.Template.Spec.NodeSelector[instanceKey]).Should(gomega.Equal(onDemandFlavor.Name))
+			gomega.Expect(k8sClient.Delete(ctx, wl1, client.PropagationPolicy(metav1.DeletePropagationBackground))).Should(gomega.Succeed())
+			framework.ExpectWorkloadsToBeAdmitted(ctx, k8sClient, cq.Name, wl2)
```
Comment (alculquicondor) [inline test/integration/scheduler/scheduler_test.go:242]: the previous test checked the flavor too

File: test/integration/scheduler/scheduler_test.go [hunk 20]
```diff
@@ -495,36 +381,23 @@ var _ = ginkgo.Describe("Scheduler", func() {
...
-				return createdJob2.Spec.Suspend
-			}, framework.Timeout, framework.Interval).Should(gomega.Equal(pointer.Bool(true)), "Job should be suspended")
+		ginkgo.It("Should admit workloads from the selected namespaces", func() {
+			ginkgo.By("checking the workloads don't get admitted at first")
+			wl1 := testing.MakeWorkload("wl1", ns.Name).Queue(queue.Name).Request(corev1.ResourceCPU, "1").Obj()
+			gomega.Expect(k8sClient.Create(ctx, wl1)).Should(gomega.Succeed())
+			framework.ExpectWorkloadsToBePending(ctx, k8sClient, wl1)
+			wl2 := testing.MakeWorkload("wl2", nsFoo.Name).Queue(queueFoo.Name).Request(corev1.ResourceCPU, "1").Obj()
+			gomega.Expect(k8sClient.Create(ctx, wl2)).Should(gomega.Succeed())
+			framework.ExpectWorkloadsToBePending(ctx, k8sClient, wl2)
```
Comment (alculquicondor) [inline test/integration/scheduler/scheduler_test.go:391]: create both workloads and do a single check for both.

```golang
framework.ExpectWorkloadsToBePending(ctx, k8sClient, wl1, wl2)
```

File: test/integration/scheduler/scheduler_test.go (hunk 19 above)
Comment (thisisprasad) [reply test/integration/scheduler/scheduler_test.go]: flavor admission check added to all workloads of the test

File: pkg/util/testing/wrappers.go [hunk 21]
```diff
@@ -156,6 +156,7 @@ func MakeWorkload(name, ns string) *WorkloadWrapper {
 								},
 							},
 						},
+						NodeSelector: map[string]string{},
```
[the remark this answers is not in the document]
Comment (thisisprasad) [reply pkg/util/testing/wrappers.go]: removed

**Comment by alculquicondor** [comment]
/approve
Please squash

**Comment by thisisprasad** [comment]
Done :)

**Comment by alculquicondor** [comment]
Thanks!
/lgtm

**Pull request merged**